File: server/warps_test.go
package main

import "testing"

func TestWarpOriginalLatitudeAndBoundaryRules(t *testing.T) {
	for _, y := range []int{355, 370, 385, 767, 782, 797} {
		for _, x := range []int{31, 609} {
			s := initial()
			s.Ball = Ball{Owner: -1, LastTouch: 7, X: float64(576-y) * terrainUnit, Z: float64(x-320) * terrainUnit, H: 1.25, FlightKind: 2, FlightStage: 2, DirX: 1, VX: 2, VZ: 3}
			s.Players[7].Stats[4] = 200
			expected := 11.2
			if x == 609 {
				expected = -11.2
			}
			if !s.sideFeature() || s.Ball.Z != expected || s.Ball.VX != 8*velocityUnit || s.Ball.VZ != 3 || s.Ball.SpeedTimer != 100 || s.Ball.FlightStage != 2 {
				t.Fatal("warp endpoints/drift", x, y, s.Ball)
			}
		}
	}
	for _, y := range []int{354, 386, 766, 798} {
		s := initial()
		s.Ball = Ball{Owner: -1, X: float64(576-y) * terrainUnit, Z: -11.3, H: 1}
		if s.sideFeature() {
			t.Fatal("outside latitude warped", y)
		}
	}
	for _, change := range []func(*Ball){func(b *Ball) { b.Z = -11.2 }, func(b *Ball) { b.Z = 11.2 }, func(b *Ball) { b.Owner = 7 }, func(b *Ball) { b.FlightKind = 2; b.FlightStage = 3 }, func(b *Ball) { b.H = 1.3 }} {
		s := initial()
		s.Ball = Ball{Owner: -1, X: 206 * terrainUnit, Z: -11.3, H: 1}
		change(&s.Ball)
		if s.sideFeature() {
			t.Fatal("ineligible warp")
		}
	}
}

File: src/warps.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import { initial } from "./game.js";
import { sideFeature } from "./features.js";
import { velocityUnit } from "./attributes.js";
const unit = 22.4 / 576;
test("warp accepts both inclusive latitude endpoints at each wall", () => {
  for (const y of [355, 370, 385, 767, 782, 797])
    for (const x of [31, 609]) {
      const s = initial();
      Object.assign(s.ball, {
        owner: -1,
        lastTouch: 7,
        x: (576 - y) * unit,
        z: (x - 320) * unit,
        h: 1.25,
        flightKind: 2,
        flightStage: 2,
        dirX: 1,
        dirZ: 0,
        vx: 2,
        vz: 3,
      });
      s.players[7].stats[4] = 200;
      assert.equal(sideFeature(s), true);
      assert.equal(s.ball.z, x === 31 ? 11.2 : -11.2);
      assert.equal(s.ball.x, (576 - y) * unit);
      assert.equal(s.ball.vx, 8 * velocityUnit);
      assert.equal(s.ball.vz, 3);
      assert.equal(s.ball.speedTimer, 100);
      assert.equal(s.ball.flightStage, 2);
      assert.equal(s.event.kind, 12);
    }
});
test("warp excludes outside latitudes, exact wall boundary, held balls and high stages", () => {
  for (const y of [354, 386, 766, 798]) {
    const s = initial();
    Object.assign(s.ball, { owner: -1, x: (576 - y) * unit, z: -11.3, h: 1 });
    assert.equal(sideFeature(s), false);
  }
  for (const change of [
    { z: -11.2 },
    { z: 11.2 },
    { z: 0 },
    { owner: 7 },
    { flightKind: 2, flightStage: 3 },
    { h: 1.3 },
  ]) {
    const s = initial();
    Object.assign(s.ball, { owner: -1, x: 206 * unit, z: -11.3, h: 1 }, change);
    assert.equal(sideFeature(s), false);
  }
});
